pkg/infrastructure/powervs/clusterapi/powervs.go

```diff
@@ -157,42 +162,48 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
 		logrus.Debugf("lbKey = %s", lbKey)
 		switch {
 		case lbExtExp.MatchString(lbKey):
-			prefix = "api."
+			if in.InstallConfig.Config.Publish == types.ExternalPublishingStrategy {
+				hostnames = append(hostnames, fmt.Sprintf("%s%s", publicPrefix, infraID))
+			}
 		case lbIntExp.MatchString(lbKey):
-			prefix = "api-int."
+			hostnames = append(hostnames, fmt.Sprintf("%s%s", privatePrefix, infraID))
+			// In the private cluster scenario, also point api.* to internal LB
+			if in.InstallConfig.Config.Publish == types.InternalPublishingStrategy {
+				hostnames = append(hostnames, fmt.Sprintf("%s%s", publicPrefix, infraID))
+			}
 		}
 		logrus.Debugf("prefix = %s", prefix)
 
-		hostname = fmt.Sprintf("%s%s", prefix, infraID)
-
-		logrus.Debugf("InfraReady: crn = %s, base domain = %s, hostname = %s, cname = %s",
-			instanceCRN,
-			in.InstallConfig.PowerVS.BaseDomain,
-			hostname,
-			*loadBalancerStatus.Hostname)
-
-		backoff := wait.Backoff{
-			Duration: 15 * time.Second,
-			Factor:   1.1,
-			Cap:      leftInContext(ctx),
-			Steps:    math.MaxInt32}
-		err = wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
-			err2 := client.CreateDNSRecord(ctx,
-				in.InstallConfig.Config.Publish,
+		for _, hostname := range hostnames {
+			logrus.Debugf("InfraReady: crn = %s, base domain = %s, hostname = %s, cname = %s",
 				instanceCRN,
 				in.InstallConfig.PowerVS.BaseDomain,
 				hostname,
 				*loadBalancerStatus.Hostname)
-			if err2 == nil {
-				return true, nil
+
+			backoff := wait.Backoff{
+				Duration: 15 * time.Second,
+				Factor:   1.1,
+				Cap:      leftInContext(ctx),
+				Steps:    math.MaxInt32}
+			err = wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
+				err2 := client.CreateDNSRecord(ctx,
+					in.InstallConfig.Config.Publish,
+					instanceCRN,
+					in.InstallConfig.PowerVS.BaseDomain,
+					hostname,
+					*loadBalancerStatus.Hostname)
+				if err2 == nil {
+					return true, nil
+				}
+				return false, err2
```

Comment:
`ExponentialBackoff` immediately returns an error if the callback returns an error. So this whole backoff call is useless here. You only need it in case you inspect the error and decide to retry the call in certain conditions.

Reply:
I believe the intent here is to wrap a function that returns only `error` to one that returns `bool, error` as `ExponentialBackoffWithContext` wants. We can alter the client functions to return `bool, error` if you'd like.. Is there something else you're suggesting?

Reply:
That's not what I mean. `ExponentialBackoff` is to be used when you want to retry a functional call (with exponential backoff). In your case, you're never retrying anything because you either stop after the first success (`return true, nil`) or you stop with the first failure (`return false, err2`). You can remove the whole `ExponentialBackoff` and just call `client.CreateDNSRecord` and you'll have the exact same behaviour.

So unless there is an error condition when you want to `return false, nil`, the `ExponentialBackoff` is useless.

Reply:
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#ExponentialBackoffWithContext
```
ExponentialBackoffWithContext repeats a condition check with exponential backoff. It immediately returns an error if the condition returns an error, [...]. If an error is returned by the condition the backoff stops immediately.
```

Reply:
ahhhhh I had thought it would retry until returning true

Reply:
you indeed are right - I've removed the backoff bits.